src/libltfs/ltfs_internal.c

```diff
@@ -982,6 +985,48 @@ int _ltfs_make_lost_found(tape_block_t ip_eod, tape_block_t dp_eod,
 	return 0;
 }
 
+/**
+ * Find the appropriate append address for a partition at which calling ltfs_write_index() will
+ * correctly restore consistency to the partition.
+ *
+ * @param dev The ltfs device to work on
+ * @param part The partition to examine
+ * @param block The starting block of the final index on the partition
+ * @return 0 if eod is the appropriate append address, >0 for the absolute block address to append
+ *         at, <0 on error
+ */
+int _ltfs_find_append_blk_after_idx(struct device_data *dev, tape_partition_t part, tape_block_t block) {
+	unsigned int n_fm_after = 0;
+	struct tc_position idx_pos;
+	struct tc_position final_fm_pos;
+	int ret = 0;
+
+	idx_pos.partition = part;
+	idx_pos.block = block;
+	check_err(tape_seek(dev, &idx_pos), 11020E, out);
+	while(ret == 0) {
+		ret = tape_spacefm(dev, 1);
+		if (ret == 0) {
+			tape_update_position(dev, &final_fm_pos);
+			n_fm_after++;
+		} else {
+			if (ret != -EDEV_EOD_DETECTED) {
+				goto out;
+			}
+		}
+	}
+	ret = 0;
+	if (n_fm_after == 2 ) {
+		/* (index | data | ... eod) - unexpected fm after data, possible incomplete index */
```

Comment:
My thinking here is that `...` could only be data blocks erroneously written by a bad implementation or an incomplete index that cannot be parsed. In either case the proper way to make the partition consistent again would be the same as if we ended in data blocks or data + fm - we need to append the most recent index after the data blocks.

Reply:
In case 2, appending the index before eod will break the index traversal mechanism, as we'll have an illegal partition layout:
`| index | data | ... | index | eod`
or an index document that will not parse because it is prefixed with unknown data:
`| index | data | ... index | eod`
I think the only sensible course of action is to assert that `...` must have been an incomplete index and overwrite it.

Reply:
> In case 2, appending the index before eod will break the index traversal mechanism, as we'll have an illegal partition layout:
> `| index | data | ... | index | eod`

I think it is a gray zone of the spec. The spec doesn't say an index needs to be placed after Data Extents.

> 5.3 Partition Layout
> This section describes the layout of an LTFS Partition in detail. An LTFS Partition contains a Label Construct followed by a Content Area. The Content Area contains zero or more Data Extents and Index Constructs in any order. The last construct in the Content Area of a complete partition shall be an Index Construct.

So I'm thinking we need to traverse all indexes if we have such kind of sequence.

Another consideration is the sequence below

1. LTFS writes a Data Extents
2. LTFS writes an index
3. LTFS writes a Data Extents
4. LTFS writes only first one block of an index
5. LTFS crashes in some reason

In this case we have following stream in Linux `data | index | data | (incomplete) index eod`. On the other hand, we would have a stream like `data | index | data | (incomplete) index | eod` on FreeBSD at this time because of automatic append of FM.

Reply:
So my proposal is 

1. Do not overwrite an (incomplete) index without following FM at recovery
2. Provide a fix into the index traversal logic to handle like `| index | data | ... | index | eod` and `data | index | data | (incomplete) index | index | eod` under another issue in the future

What do you think?

Reply:
> I think it is a gray zone of the spec. The spec doesn't say an index needs to be placed after Data Extents.

The spec does specifically state in section 5.3 that the content area consists of data extents and index constructs in no particular order. However, per 5.2.2 and 5.2.3, data extents by definition do not include filemarks and index constructs must consist of a beginning filemark, index blocks, and a terminating filemark. `| index | data | ... | index | eod ` would therefore be an invalid content region because `| ... |` _must_ be an index construct as defined in the spec. If the blocks `...` are not a parsable index document, then is not a valid index construct, and should not be retained when recovering the volume consistency.

Prior logic in the media check will restore the terminating filemark to `| ... eod` if `...` is a parsable index, i.e. `| index (was '...') | eod`, and the append position will just remain at eod after the filemark is restored since there are no additional filemarks after the repaired final index construct.

Reply:
> So my proposal is
> 
> 1. Do not overwrite an (incomplete) index without following FM at recovery
> 2. Provide a fix into the index traversal logic to handle like `| index | data | ... | index | eod` and `data | index | data | (incomplete) index | index | eod` under another issue in the future
> 
> What do you think?

A complete index without a following FM will not be overwritten, as the following FM will be restored and the append location will remain at EOD since that index is recognized as the most recent and fewer than two filemarks follow it.

At present there does not appear to be a way to identify that an index is incomplete. Regardless, an incomplete (unparsable) index in an index construct is a violation of the layout specification. The best case scenario for retaining  blocks from an incomplete index would be to rewrite them as a data exent and reference them in lost+found, given full_recovery is true. That gets more complicated though because libltfs would have to read and store those blocks on external storage before rewriting them as a data extent. I think it's more prudent just to toss the incomplete index (which functionally is the same as no index) and let the unreferenced data extent be saved in lost+found if full_recovery is true.

Reply:
My explanation might be not clear. Let me explain it again.

The conditions the CAM backed can create because of the automatic FM append feature are

1. `| index | data | eod`
2. `| index | data | (incomplete) index | eod`

In my understanding, the fixed code might recover these data stream like 

i. `| index | data | index | eod`
ii. `-LTFS_OP_TO_INV` because `n_fm_after == 3`

Is it correct?

If so, I would like to make the condition below an error as same as case 2 because the point is as same as case2. And improve it in the new issue. 

3. `| index | data | (incomplete) index eod`

The condition can be created only by the low SCSI handling backends like the sg backend.

Reply:
> My original thinking was that -LTFS_OP_TO_INV is not appropriate in case 3 because we could return consistency to the partition by laying down the most recent index after that data extent, assuming that the index write had been interrupted. However, since it is functionally identical to case 2 above there's reasonable expectation that both would be handled the same way. For both cases, without examining the blocks in the incomplete index constructs we cannot know for certain if we have an incomplete index or a poorly behaving implementation.

Honestly, I just have realized these conditions (case 1, 2 and 3) in this discussion. And I read through the spec again. In my sense, the spec shall state how to recover the tape. I believe we don't want to overwrite any data on the tape for safety of the customer's data. Really it is rare case but I have a few experience that they save me from difficult situations. 

> ltfsck (and ltfs_check_medium() by proxy) needs to be capable of restoring a volume in these conditions at least to a state that the coherency can be validated when ltfs_check_medium() is called with fix == true. Maybe ltfs_check_medium() should always restore consistency by truncating after the final identified data extent by writing an index, and only return -LTFS_OP_TO_INV in these cases you have identified if fix == false. Does that seem reasonable?

It's a not permanent solution but a short time solution (work around) for now. We need to define how to recover these conditions in a new issue. 

Could you help me for this topic? I will inform you once I created a new issue about recovering (and traversing after recovery) these case.

Reply:
> It's a not permanent solution but a short time solution (work around) for now. We need to define how to recover these conditions in a new issue.

Ok, if that is acceptable for the immediate term I will make that change and update the PR.

> Could you help me for this topic? I will inform you once I created a new issue about recovering (and traversing after recovery) these case.

Sure, I would be happy to help in whatever way I can!

Reply:
One more note - for future refinements it would be very helpful to plumb a forward-traversing rollback option through ltfsck down to libltfs. In the case that we end up with an artifact on tape that prevents the backward traversal from being able to locate all index generations, it would be beneficial have the ability to fall back on a forward traversing rollback that can locate the desired generation.

Reply:
> One more note - for future refinements it would be very helpful to plumb a forward-traversing rollback option through ltfsck down to libltfs. In the case that we end up with an artifact on tape that prevents the backward traversal from being able to locate all index generations, it would be beneficial have the ability to fall back on a forward traversing rollback that can locate the desired generation.

Sorry I can't understand what you are thinking correctly. 

At this time, `ltfs_traverse_index_backward()` and `ltfs_traverse_index_forward()` is implemented into libltfs. Could you open a new issue about this and provide a clear statement?

Reply:
Sure, I can do that. What I'm saying here is that ltfsck (and the maybe underlying rollback function, if I recall correctly) can list rollback points via forward traversal, but does not have the option to perform a rollback to a generation that is found with the forward traversal. As a result, if there is an anomoly between generation N and N+M that causes the backward traversal to error out before reaching generation N, ltfsck cannot perform that rollback. The rollback action for ltfsck should honor the -v/--travese optarg.

Reply:
> What I'm saying here is that ltfsck can list rollback points via forward traversal

I decided this behavior intentionally before. But I can't remind why it is not a good behavior at all. Another behavior change might make it not a problem anymore..